### PR TITLE
[cpp-pistache]Add support for map

### LINKED ISSUE #745: [C++][Pistache-server] Wrong "toJson" and "fromJson" method generation - compilation failed
##### Description

The C++ generated code does not compile because of the following errors: 
 1) there is no matching function for ModelBase::toJson taking as argument std::map<std::string, std::vector<Point2D>>
2) the generated code is calling "fromJson" method as it is a member of std::map type
3) the generated code is calling a set method that is not generated

These error occur in model/Detections.cpp file.

 
Detection.h:
```
#include "ModelBase.h"

#include "BoundingRectangle.h"
#include <string>
#include "Point2D.h"
#include <map>
#include <vector>

namespace org {
namespace openapitools {
namespace server {
namespace model {

class  Detection
    : public ModelBase
{
public:
    Detection();
    virtual ~Detection();

    nlohmann::json toJson() const override;
    void fromJson(nlohmann::json& json) override;
 
    std::map<std::string, std::vector<Point2D>>& getKeyPoints();
    //setKeyPoints is missing
    bool keyPointsIsSet() const;
    void unsetKeyPoints();


protected:
   
    std::map<std::string, std::vector<Point2D>> m_KeyPoints;
    bool m_KeyPointsIsSet;
 
};

}
}
}
}

#endif /* Detection_H_ */
```
Detection.cpp:
```
nlohmann::json Detection::toJson() const
{
    //
    //
    if(m_KeyPointsIsSet)
    {
          //this call
        val["keyPoints"] = ModelBase::toJson(m_KeyPoints);
    }
   //
   //
    return val;
}

void Detection::fromJson(nlohmann::json& val)
{
    //
    //
    if(val.find("keyPoints") != val.end())
    {
        if(!val["keyPoints"].is_null())
        {
            std::map<std::string, std::vector<Point2D>> newItem;
            //and these 2 lines below
            newItem.fromJson(val["keyPoints"]);
            setKeyPoints( newItem );
        }
        
    }
}
```

##### openapi-generator version
3.2.0 (I cloned the latest master branch 2 days ago)

##### OpenAPI declaration file content or url
https://raw.githubusercontent.com/deepviss/deepviss-server-example/master/src/main/resources/deep-viss.json

##### Command line used for generation
Inside cloned repository, after mvn clean package:
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate /
      -i deep-viss.json /
      -g cpp-pistache-server /
       -o ../openapi-pistache-server

##### Steps to reproduce
  Note: Assuming Pistache is installed
 1) Go to output folder
 2) copy json.hpp (from https://github.com/nlohmann/json) to model folder
 3) mkdir build
 4) cd build
 5) cmake ..
 6) make

### PR DESCRIPTION
**Pull request by etherealjoy**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.4.x`, `4.0.x`. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Add support for map
Add support for nested maps
Simplify Array and Map Helper

Fixes #745 

@stkrwork @MartinDelille @fvarose @ravinikam

### CODE REVIEW

**Comment by MartinDelille** [comment]
Shouldn't map support be handled at the AbstractCppCodegen level?

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 1]
```diff
@@ -39,100 +39,85 @@ public:
...
+private:
+    template<typename T, typename std::enable_if<!std::is_base_of<ModelBase, T>::value>::value>
+    static void itemFromJson(T &item, nlohmann::json &json){
+        item = json;
+    }
+    static void itemFromJson(ModelBase &item, nlohmann::json &json){
+        item.fromJson(json);
+    }
+    template<typename T, typename std::enable_if<!std::is_base_of<ModelBase, T>::value>::value>
+    static nlohmann::json itemtoJson(T &item){
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: is the object going to change here? if not make the parameter const.

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 2]
```diff
@@ -39,100 +39,85 @@ public:
...
+        return item;
+    }
+    static nlohmann::json itemtoJson(ModelBase &item){
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: same as above

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 3]
```diff
@@ -39,100 +39,85 @@ public:
...
+            for (auto &item : json.items()) {
+                T entry;
+                itemFromJson(entry, item.value());
+                val.push_back(entry);
+            }
+        }
+        return val;
+    }
+    template<typename T>
+    static nlohmann::json toJson(std::vector<T> val) {
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: why not const&?

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 4]
```diff
@@ -39,100 +39,85 @@ public:
...
+private:
+    template<typename T, typename std::enable_if<!std::is_base_of<ModelBase, T>::value>::value>
+    static void itemFromJson(T &item, nlohmann::json &json){
+        item = json;
+    }
+    static void itemFromJson(ModelBase &item, nlohmann::json &json){
+        item.fromJson(json);
+    }
+    template<typename T, typename std::enable_if<!std::is_base_of<ModelBase, T>::value>::value>
+    static nlohmann::json itemtoJson(T &item){
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: same as above

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 5]
```diff
@@ -39,100 +39,85 @@ public:
...
+        return item;
+    }
+    static nlohmann::json itemtoJson(ModelBase &item){
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: same as above

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 6]
```diff
@@ -39,100 +39,85 @@ public:
...
+                T entry;
+                itemfromJson(entry, item.value());
+                val.insert(val.end(),
+                        std::pair<std::string, T>(item.key(), entry));
+            }
+        }
+        return val;
+    }
+    template<typename T>
+    static nlohmann::json toJson(std::map<std::string, T> val) {
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: same as above

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache [hunk 7]
```diff
@@ -39,100 +39,85 @@ public:
...
-	static nlohmann::json toJson(std::map<std::string, T> val) {
-		nlohmann::json json;
-		for (auto const& item : val) {
-		  nlohmann::json jitem = item.second;
-		  json[item.first] = jitem;
-		}
-		return json;
-	}
+    template<typename T>
+    static std::map<std::string, T> & fromJson(nlohmann::json& json) {
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: same as above

**Comment by etherealjoy** [comment]
@stkrwork 
I updated to use const & where possible.
The `const &` is read `right to left` so we can put the `const`  before or after the type

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 1 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 2 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 3 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 4 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 6 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache (hunk 7 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/modelbase-header.mustache]: Done

**Comment by etherealjoy** [comment]
Many thanks for the review.

**Pull request merged**